Sources/NIOSSL/SSLCertificate.swift

```diff
@@ -407,73 +397,108 @@ extension NIOSSLCertificate: Hashable {
     }
 }
 
-/// A helper sequence object that enables us to represent subject alternative names
-/// as an iterable Swift sequence.
-internal class SubjectAltNameSequence: Sequence, IteratorProtocol {
-    typealias Element = NIOSSLCertificate.AlternativeName
+public struct _SubjectAlternativeName {
+    public var nameType: NameType
+    public var contents: Contents
 
-    private let nameStack: OpaquePointer
-    private var nextIdx: Int
-    private let stackSize: Int
+    public struct NameType: Hashable {
+        var rawValue: Int
 
-    init(nameStack: OpaquePointer) {
-        self.nameStack = nameStack
-        self.stackSize = CNIOBoringSSLShims_sk_GENERAL_NAME_num(nameStack)
-        self.nextIdx = 0
+        fileprivate init(_ rawCode: Int) {
+            self.rawValue = rawCode
+        }
+
+        fileprivate init(_ rawCode: Int32) {
+            self.init(Int(rawCode))
+        }
+        
+        public static let email = Self(GEN_EMAIL)
+        public static let dnsName = Self(GEN_DNS)
+        public static let ipAddress = Self(GEN_IPADD)
+    }
+
+    public struct Contents {
+        // only part of this type to keep a strong reference to the underlying storage of `buffer`
+        private let collection: _SubjectAlternativeNames
+        // lifetime automatically managed by `collection`
+        fileprivate let buffer: UnsafeBufferPointer<UInt8>
+        
+        internal init(collection: _SubjectAlternativeNames, buffer: UnsafeBufferPointer<UInt8>) {
+            self.collection = collection
+            self.buffer = buffer
+        }
     }
+    
+    internal enum IPAddress {
+        case ipv4(in_addr)
+        case ipv6(in6_addr)
+    }
+}
 
-    private func addressFromBytes(bytes: UnsafeBufferPointer<UInt8>) -> NIOSSLCertificate.IPAddress? {
-        switch bytes.count {
+extension _SubjectAlternativeName.Contents: RandomAccessCollection {
+    public var startIndex: Int { buffer.startIndex }
+    public var endIndex: Int { buffer.endIndex }
+    
+    public subscript(position: Int) -> UInt8 {
+        get {
+            buffer[position]
+        }
+    }
+}
+
+extension _SubjectAlternativeName.IPAddress {
+    init?(_ subjectAlternativeName: _SubjectAlternativeName) {
+        guard subjectAlternativeName.nameType == .ipAddress else {
+            return nil
+        }
+        switch subjectAlternativeName.contents.count {
         case 4:
-            let addr = bytes.baseAddress?.withMemoryRebound(to: in_addr.self, capacity: 1) {
+            let addr = subjectAlternativeName.contents.buffer.baseAddress?.withMemoryRebound(to: in_addr.self, capacity: 1) {
```

Comment:
Instead of reaching to `buffer` directly, can we use a `withUnsafePointer` function? It's possible to drop the reference to `_SubjectAlternativeNames` when we use a direct member access like this.